dbrequests/database.py

```diff
@@ -9,43 +10,78 @@
+      in send_data.
+    - escape_percentage: (bool) escape percentages when reading queries from a
+      file.
+    - remove_comments: (bool) remove comments when reading queries from a file.
+    - kwargs:
+        - creds: (dict) deprecated, provide a dict as db_url
+        - ...: all arguments are passed to sqlalchemy.create_engine
     """
 
-    def __init__(self, db_url=None, creds=None, sql_dir=None, connection_class=DefaultConnection,
+    def __init__(self, db_url=None, sql_dir=None,
+                 connection_class=DefaultConnection,
                  escape_percentage=False, remove_comments=False, **kwargs):
-        # If no db_url was provided, fallback to $DATABASE_URL or creds.
-        self.db_url = db_url or os.environ.get('DATABASE_URL')
-        self.sql_dir = sql_dir or os.getcwd()
-        if not self.db_url:
-            try:
-                user = creds['user']
-                password = creds['password']
-                host = creds['host']
-                db = creds['db']
-                dialect = creds.get('dialect', 'mysql')
-                driver = creds.get('driver', 'pymysql')
-                self.db_url = '{}+{}://{}:{}@{}/{}'.format(
-                    dialect, driver, user, password, host, db)
-            except:
-                raise ValueError('You must provide a db_url or proper creds.')
+        if db_url is None:
+            db_url = os.environ.get('DATABASE_URL')
+            if db_url is None:
+                db_url = kwargs.pop('creds', None)
+                if db_url:
+                    warnings.warn(
+                        "Parameter 'creds' is depreacated in favor of db_url.",
+                        DeprecationWarning)
+                else:
+                    raise ValueError('db_url is missing')
+        if isinstance(db_url, str):
+            self.db_url = db_url
+        elif isinstance(db_url, dict):
+            db_url = db_url.copy()
+            self.db_url = '{}+{}://{}:{}@{}:{}/{}'.format(
+                db_url.pop('dialect', 'mysql'),
+                db_url.pop('driver', 'pymysql'),
+                db_url.pop('user'),
+                db_url.pop('password'),
+                db_url.pop('host', '127.0.0.1'),
+                db_url.pop('port', 3306),
+                db_url.pop('db'))
+            connect_args = kwargs.pop('connect_args', {})
+            connect_args.update(db_url)
+            if len(connect_args):
+                kwargs['connect_args'] = connect_args
+        else:
+            raise ValueError('db_url has to be a str or dict')
 
+        self.sql_dir = sql_dir or os.getcwd()
         self._escape_percentage = escape_percentage
         self._remove_comments = remove_comments
-        self._engine = create_engine(self.db_url, **kwargs)
+        self._engine = self._init_engine(self.db_url, **kwargs)
         self._open = True
         self.connection_class = connection_class
 
+    def _init_engine(self, url, **kwargs):
```

Comment:
whats the purpose of this? and shouldnt it be a static method?

Reply:
The purpose here is, so that I can override it in a subclass. I need some way to pass down default settings in `connect_args`. Previously I was doing that by overriding init. Now this approach is not feasible because I need the handling of the db_url. It's a bit twisted, but within the OO paradigm I am not sure what else to do...

Reply:
Also the connect_args depend on the driver.

Reply:
I see. We could also move the construction of db_url out of the __init__ method as a static method. Either way, if self is not needed in the method pls use the staticmethod decorator.

Reply:
- _init_engine: self is needed in the derived subclass. Also I would like to be able to call super() in the derived class, so that upstream changes are picked up automatically. Having a static method will complicate both things, and requires me to change the complete setup of the derived class. I refactored the method, so that it now needs self.
- _init_db_url: In the init block we mostly generate the db_url, but also extract the connect_args from the credentials object. These are then added to the kwargs before passed to create_engine. I do not believe that having a static method would make this process much simpler.

If you have a suggestion how to make the init process more modular, then I am all ear for it. I have already spent quite some time to deal with this. In the current OO setup I do not know how to improve it without a complete rewrite of the Database interface; and breaking existing code along the way.

Reply:
btw: I refactored the init method, that it will now call _init_db_url. This method will now do two things: (a) init db_url (b) return an updated kwargs object.